packages/medusa/src/services/cart.ts

```diff
@@ -666,185 +716,180 @@ class CartService extends BaseService {
+            const customer = await this.createOrFetchUserFromEmail_(data.email)
+            cart.customer = customer
+            cart.customer_id = customer.id
+            cart.email = customer.email
+          }
         }
-      }
 
-      if (typeof update.region_id !== "undefined") {
-        const countryCode =
-          (update.country_code || update.shipping_address?.country_code) ?? null
-        await this.setRegion_(cart, update.region_id, countryCode)
-      }
-
-      if (
-        typeof update.customer_id !== "undefined" ||
-        typeof update.region_id !== "undefined"
-      ) {
-        await this.updateUnitPrices(cart, update.region_id, update.customer_id)
-      }
-
-      const addrRepo = manager.getCustomRepository(this.addressRepository_)
-      if ("shipping_address_id" in update || "shipping_address" in update) {
-        let address: string | Partial<Address> | undefined
-        if (typeof update.shipping_address_id !== "undefined") {
-          address = update.shipping_address_id
-        } else if (typeof update.shipping_address !== "undefined") {
-          address = update.shipping_address
+        if (typeof data.region_id !== "undefined") {
+          const countryCode =
+            (data.country_code || data.shipping_address?.country_code) ?? null
+          await this.setRegion_(cart, data.region_id, countryCode)
         }
 
-        if (typeof address !== "undefined") {
-          await this.updateShippingAddress_(cart, address, addrRepo)
+        if (
+          typeof data.customer_id !== "undefined" ||
+          typeof data.region_id !== "undefined"
+        ) {
+          await this.updateUnitPrices_(cart, data.region_id, data.customer_id)
         }
-      }
 
-      if ("billing_address_id" in update || "billing_address" in update) {
-        let address: string | Partial<Address> | undefined
-        if (typeof update.billing_address_id !== "undefined") {
-          address = update.billing_address_id
-        } else if (typeof update.billing_address !== "undefined") {
-          address = update.billing_address
+        const addrRepo = transactionManager.getCustomRepository(
+          this.addressRepository_
+        )
+
+        const billingAddress = data.billing_address_id ?? data.billing_address
+        if (billingAddress !== undefined) {
+          await this.updateBillingAddress_(cart, billingAddress, addrRepo)
         }
 
-        if (typeof address !== "undefined") {
-          await this.updateBillingAddress_(cart, address, addrRepo)
+        const shippingAddress =
+          data.shipping_address_id ?? data.shipping_address
+        if (shippingAddress !== undefined) {
+          await this.updateShippingAddress_(cart, shippingAddress, addrRepo)
         }
-      }
 
-      if (typeof update.discounts !== "undefined") {
-        const previousDiscounts = cart.discounts
-        cart.discounts = []
+        if (data.discounts?.length) {
```

Comment:
Hey @adrien2p , it seems like because of this line, we are not able to remove discount from a cart.